### ISSUE TITLE
Bug in hammer projection with lon_0>0

### ISSUE THREAD

**Issue by William-gregory** [matplotlib/basemap#463]
When using Basemap 'hammer' projection with lon_0 set to anything greater than 0 degrees, the map is plotted as the mirror image of reality. Setting lon_0 to negative values (like -180) works fine, but I don't want to do this as then none of my data would fit this projection and I would have to re-grid everything.

I've subsequently uninstalled & reinstalled anaconda3 (now have python 3.7) and have also reinstalled Basemap (v.1.2.0), and am still getting the same problem.

Here is the code:

import matplotlib as mpl
from mpl_toolkits.basemap import Basemap
import matplotlib.pyplot as plt

m = Basemap(projection='hammer', resolution = 'l', lon_0=180)

fig = plt.figure(figsize=(12,12))
m.drawcoastlines(linewidth=0.3)
m.fillcontinents(color='lightgrey',lake_color='white')
m.drawmapboundary(fill_color='white')
plt.show()

![hammer_proj](https://user-images.githubusercontent.com/37665577/59605438-b68c7380-9106-11e9-9cb4-ab9555dfdf70.png)

**Comment by ghzuo** [comment]
I have the same problem in Basemap 'robin' projection. I had updated the basemap to 1.2.1 and matplotlib 3.1. 

And I think there is another problem is relative to this problem. When lon_0 is not equal 0, all the background functions, like drawlsmask(), bluemarble(), ..., are useless.

**Comment by pochedls** [comment]
Same issue with eck4 using basemap 1.2.0, matplotlib 3.0.3, python 3.7.1. 

@William-gregory / @ghzuo - were you able to use a different version to circumvent this issue?

**Comment by WeatherGod** [comment]
Most likely, the issue is related to pyproj4 or proj4 as most of the work
for transforms are off-loaded to proj4.

**Comment by gkvallis** [comment]
I also find that, in order for the code to even run (in an anaconda environment), I need to preface it with:

import os
import conda
conda_file_dir = conda.__file__
conda_dir = conda_file_dir.split('lib')[0]
proj_lib = os.path.join(os.path.join(conda_dir, 'share'), 'proj')
os.environ["PROJ_LIB"] = proj_lib

Previously, this preface was not needed. I do not know whether this is related to the bug the OP reported. The inverted images occur with many of the map projections, but not for 'cyl', which is just a lat-long projection.

**Comment by sbonaime** [comment]
Any idea to solve this strange problem ?

**Comment by pochedls** [comment]
@sbonaime - I have largely transitioned to cartopy, but I recall using a negative value, e.g., `lon_0=-180`, gave me something that looked right.

**Comment by sbonaime** [comment]
@pochedls I should update my old script !
with lon_0=0, my map is ok but I need lon_0=15...
Any workaround like installling an "old" lib ?

**Comment by pochedls** [comment]
@sbonaime - I assume it would be possible to address this with an older library (perhaps just creating a conda environment with an older version of basemap would work). `lon_0=-345` might work?

**Comment by sbonaime** [comment]
@pochedls lon_0=-345 is not working... Maybe I should take some time to update to cartopy. Thanks

**Comment by PhillCli** [comment]
So, I've hit a similar issue, code I've used to reproduce, blue-marble with robin projection

```python
from mpl_toolkits.basemap import Basemap
import matplotlib.pyplot as plt

m = Basemap(projection="robin", lon_0=0, resolution=None)
m.bluemarble()
plt.show()
```
Expected output:
![image](https://user-images.githubusercontent.com/32617178/185132728-5be63d7c-d923-402b-a39e-ab055fb0edfe.png)

It works on all tested base map versions - 1.1.0 - 1.3.4 only if I downgrade matplotlib to `3.4.3`, `3.5.X` release doesn't work with any version of base map I've tried.

TL;DR try downgrading matplotlib, ~3.4.3 worked for me

**Comment by molinav** [comment]
@PhillCli Thanks for the feedback and the hint about the `matplotlib` version. Do you have the full traceback that you get when using `matplotlib` 3.5.x? Probably some deprecated `matplotlib` syntax was removed in 3.5.x and `basemap` is still using the old syntax. I will take a look to it when I get some free time.

**Comment by PhillCli** [comment]
> Do you have the full traceback that you get when using matplotlib 3.5.x?

It's a silent failure, the plot is generated but is completely empty except for an outline. Just like in this post [comment](https://github.com/matplotlib/basemap/issues/463#issuecomment-723505972)

**Comment by molinav** [comment]
@PhillCli Removing the `projection` argument is equivalent to setting `projection="cyl"`.

From your code snippet, if you try to do the following with the Robin projection:
```
plt.imshow(m._bm_rgba_warped)
```
you can see that the transformed image is actually there and it is transformed correctly (note that the image will appear upside down in the figure, but this also occurs in `matplotlib` 3.4.x, so I assume it is the normal behaviour). When setting `projection="cyl"` (or not defining it), the underlying image is stored in a different hidden attribute:
```
plt.imshow(m._bm_rgba)
```

So the problem does not seem to be with the projection transformations but with the rendering of the image by `matplotlib`. Furthermore, if I just add one line to your code snippet then the image is shown (using `matplotlib` version 3.5.3):
```
from mpl_toolkits.basemap import Basemap
import matplotlib.pyplot as plt

m = Basemap(projection="robin", lon_0=0, resolution=None)
im1 = m.bluemarble()
im1.axes.add_image(im1)
plt.show()
```
> ![bluemarble_robin](https://user-images.githubusercontent.com/9979942/185264486-ec36defe-0572-44fc-b584-26751783b513.png)

The problem that you described propagates to all the `Basemap` methods related to drawing image backgrounds, since all of them rely on `Basemap.warpimage`, which in the end is calling `Basemap.imshow` to show the image.

**Comment by molinav** [comment]
Going a bit more into the `im1` object from my line added, just a couple of outputs from an interactive session:
```
In [24]: type(im1.axes)
Out[24]: matplotlib.axes._subplots.AxesSubplot

In [25]: im1.axes.get_images()[0]
Out[25]: <matplotlib.image.AxesImage at 0x204343eaaf0>

In [26]: im1.axes.get_images()[0] is im1
Out[26]: True
```
so `im1.axes` actually has a list of images containing only `im1`, which is our rebel axes image. At least this looks normal to me.

**Comment by PhillCli** [comment]
@molinav Thanks for the investigation, I'm not that familiar with the internals of Basemap.

Good to know the image is still properly transformed, and just a matter of explicitly adding it to Axes object.

That would make sense that something about rendering changed in 3.5.3 that breaks the default Basemap imshow, wonder if that was also the case for the @sbonaime issue.

It might be valuable to update the docs I think that's where the snippet I tried to fix originated from [here](https://matplotlib.org/basemap/users/geography.html).

**Comment by YilongWang** [comment]
I have the similar issue with "moll" projection, and find a very rough solution. See if the steps described in #577 can solve your problem. 
> When using Basemap 'hammer' projection with lon_0 set to anything greater than 0 degrees, the map is plotted as the mirror image of reality. Setting lon_0 to negative values (like -180) works fine, but I don't want to do this as then none of my data would fit this projection and I would have to re-grid everything.
> 
> I've subsequently uninstalled & reinstalled anaconda3 (now have python 3.7) and have also reinstalled Basemap (v.1.2.0), and am still getting the same problem.
> 
> Here is the code:
> 
> import matplotlib as mpl from mpl_toolkits.basemap import Basemap import matplotlib.pyplot as plt
> 
> m = Basemap(projection='hammer', resolution = 'l', lon_0=180)
> 
> fig = plt.figure(figsize=(12,12)) m.drawcoastlines(linewidth=0.3) m.fillcontinents(color='lightgrey',lake_color='white') m.drawmapboundary(fill_color='white') plt.show()
> 
> ![hammer_proj](https://user-images.githubusercontent.com/37665577/59605438-b68c7380-9106-11e9-9cb4-ab9555dfdf70.png)

**Comment by YilongWang** [comment]
I have the similar issue with "moll" projection, and find a very rough solution. See if the steps described in #577 can solve your problem. 
                # draw another filled patch, with no boundary.
                limb2 = self.drawmapboundary(linewidth=0, ax=ax**,fill_color='none'**)
                self._mapboundarydrawn = limb2
If seems limb2 is not necessary in current version of basemap.
> @molinav Thanks for the investigation, I'm not that familiar with the internals of Basemap.
> 
> Good to know the image is still properly transformed, and just a matter of explicitly adding it to Axes object.
> 
> That would make sense that something about rendering changed in 3.5.3 that breaks the default Basemap imshow, wonder if that was also the case for the @sbonaime issue.
> 
> It might be valuable to update the docs I think that's where the snippet I tried to fix originated from [here](https://matplotlib.org/basemap/users/geography.html).

**Comment by molinav** [comment]
Tracing back with Python 3.7, the failure related to the flipped coastlines occurs in the transition of the `pyproj` dependency from version 1.9.6 into 2.0.0, when `pyproj` migrated from PROJ 4.x into PROJ 6.0. Something changed between these two PROJ versions.

**Comment by molinav** [comment]
I close this issue as complete after applying the patch proposed by @YilongWang. If any problem persists, please free to reopen this issue or to create a new one.
